corehq/apps/locations/models.py

```diff
@@ -274,6 +274,13 @@ def couch_locations(self, wrapped=True):
             return itertools.imap(Location.wrap, locations)
         return locations
 
+    def accessible_to_user(self, domain, user):
+        if user.has_permission(domain, 'access_all_locations'):
+            return self
+
+        users_location = user.get_sql_location(domain)
+        return self & users_location.get_descendants(include_self=True)
```

Comment:
I'd like to see some tests for this, since it's a pretty complex chain of filters.